File: src/cii_platform/services/applicability.py
"""CII 적용 대상 판정의 단일 출처 (#653).

``API_SPEC §2.3``이 정한 판정(``gross_tonnage >= 5,000``)은 등록·수정 시점에 서버가
내려 ``vessel.is_cii_applicable_hint``에 굳는다. 이 모듈은 그 판정을 **계산 결과와
문서가 같은 말로 옮기게** 하는 자리다.

## 왜 따로 두는가

판정 규칙이 `services/vessel.py`(등록 시 산정)와 `services/voyage_cii.py`(경고 산정)에
각각 있었고, 임계값 상수(``CII_APPLICABLE_GT_THRESHOLD``)도 두 파일에 따로 있었다.
같은 규칙이 두 곳에 있으면 한쪽만 고쳐진다 — `#653`이 드러낸 것도 결국 **판정은 있는데
그 결과를 옮기는 곳이 한 군데뿐**이라는 문제였다.

## 세 상태다 — 둘이 아니다

``is_cii_applicable_hint``는 boolean이라 「해당/미해당」 둘로 보이지만, **미해당의
원인이 둘**이다.

- ``NOT_APPLICABLE`` — GT를 알고 그것이 5,000 미만이다. 판정이 끝난 상태다
- ``UNKNOWN`` — GT가 ``NULL``이다. **판정할 근거가 없다.** 「적용 대상이 아니다」라고
  단정하면 사용자가 확인된 사실로 읽는다

둘을 합치면 **총톤수를 넣지 않은 사용자가 「이 배는 규제 대상이 아니다」로 읽는다.**
데모 시드의 실선 2척(``STAR SKIPPER`` · ``DONGJIN ENDURANCE``)이 정확히 후자이며,
`#587`의 제원 조사 회신을 기다리는 중이다.
"""

from __future__ import annotations

from decimal import Decimal

#: ``API_SPEC §2.3`` · ``DB_SCHEMA §2.1`` — ``is_cii_applicable_hint`` 자동 산정 기준.
#: ``PRD §3.1``의 MARPOL Annex VI Reg.28 적용 하한과 같은 값이며, 그 정본은 IMO 원문이다.
CII_APPLICABLE_GT_THRESHOLD = Decimal("5000")

#: ``API_SPEC §1.6`` — GT를 알고 그것이 5,000 미만.
WARNING_NON_CII_VESSEL = "NON_CII_VESSEL"

#: ``API_SPEC §1.6`` — GT가 없어 판정 자체가 불가 (`#653` 신설).
WARNING_CII_APPLICABILITY_UNKNOWN = "CII_APPLICABILITY_UNKNOWN"

#: 판정 3상태. 문자열로 두는 것은 응답에 그대로 실릴 수 있게 하기 위함이 아니라,
#: 서비스 사이에서 뜻이 드러나는 이름으로 오가게 하기 위함이다.
STATE_APPLICABLE = "APPLICABLE"
STATE_NOT_APPLICABLE = "NOT_APPLICABLE"
STATE_UNKNOWN = "UNKNOWN"


def applicability_state(gross_tonnage: Decimal | None) -> str:
    """총톤수로 CII 적용 대상 3상태를 판정한다.

    저장된 ``is_cii_applicable_hint``가 아니라 **GT 원본**을 본다. 힌트는 boolean이라
    ``UNKNOWN``을 표현할 수 없고, 등록 시점에 굳은 값이라 제원이 그 뒤에 바뀌었으면
    현재 GT와 어긋날 수 있다. 판정 근거를 매번 원본에서 다시 읽는다.
    """
    if gross_tonnage is None:
        return STATE_UNKNOWN
    if Decimal(gross_tonnage) < CII_APPLICABLE_GT_THRESHOLD:
        return STATE_NOT_APPLICABLE
    return STATE_APPLICABLE


def applicability_warnings(vessel) -> list[str]:
    """선박 하나에 붙는 CII 적용 대상 경고 (``API_SPEC §1.6``).

    적용 대상이면 빈 목록이다 — **정상 상태에 경고를 붙이지 않는다.**

    호출부가 이 결과를 자기 경고 목록에 이어 붙인다. 계산 결과(기능①·②)·YTD·실시간
    CII·리포트 2종이 모두 같은 함수를 거치므로, 한 화면에만 경고가 붙고 다른 화면은
    조용한 상태가 생기지 않는다.
    """
    state = applicability_state(vessel.gross_tonnage)
    if state == STATE_NOT_APPLICABLE:
        return [WARNING_NON_CII_VESSEL]
    if state == STATE_UNKNOWN:
        return [WARNING_CII_APPLICABILITY_UNKNOWN]
    return []

File: frontend/src/components/applicability.ts
/**
 * CII 적용 대상 판정을 화면 언어로 옮기는 단일 규칙 (#653).
 *
 * ## 왜 한 곳인가
 *
 * 종전에는 이 규칙이 `vessel-registration/resultRules.ts` 안에만 있었고, 그 화면은
 * **등록 직후 한 번만 지나간다.** 이후 목록·상세·대시보드 어디에서도 그 선박이
 * 규제 대상인지 알 수 없었다 — 계산은 정상적으로 도는데 그 값이 규제상 무의미할 수
 * 있다는 사실을 화면이 말하지 않았다.
 *
 * ## 화면이 다시 판정하지 않는다
 *
 * 임계값(GT 5,000)을 이 파일에 적지 않는다. 판정은 서버 소관이고(`API_SPEC §2.3`),
 * 숫자를 화면에 박으면 규제 하한이 바뀔 때 화면만 낡는다. 화면은 서버가 내린
 * `is_cii_applicable_hint`를 **그대로 받아** 「미해당의 원인」만 가른다.
 */

/**
 * 판정 3상태. 서버 `services/applicability.py`의 상태명과 같은 이름을 쓴다 —
 * 다르게 부르면 로그와 화면을 대조할 때 매번 번역해야 한다.
 */
export type ApplicabilityState = 'APPLICABLE' | 'NOT_APPLICABLE' | 'UNKNOWN'

export interface ApplicabilityInput {
  /** 서버 판정 (`API_SPEC §2.3`). 화면이 만들지 않는다. */
  isCiiApplicableHint: boolean
  /**
   * 총톤수. **판정에 쓰지 않고 「미해당의 원인」을 가르는 데만 쓴다.**
   * 응답에 따라 number(`§2.1`)이거나 문자열이라 둘 다 받는다.
   */
  grossTonnage: number | string | null
}

/**
 * 「미해당」의 두 원인을 가른다.
 *
 * 서버는 GT 기준으로 판정하는데, **GT가 비어 있어도 `false`가 나온다.** 두 경우에
 * 같은 말을 하면 총톤수를 넣지 않은 사용자가 「이 배는 규제 대상이 아니다」로 읽는다.
 * 데모 시드의 실선 2척(`STAR SKIPPER` · `DONGJIN ENDURANCE`)이 정확히 후자다.
 */
export function applicabilityState({
  isCiiApplicableHint,
  grossTonnage,
}: ApplicabilityInput): ApplicabilityState {
  if (isCiiApplicableHint) return 'APPLICABLE'
  if (grossTonnage === null) return 'UNKNOWN'
  return 'NOT_APPLICABLE'
}

/**
 * 배지에 적는 짧은 라벨. 전체 문구는 {@link APPLICABILITY_FULL_TEXT}가 갖는다.
 *
 * 목록에서 스무 척을 훑을 때 한 줄이 두 줄로 접히면 표가 무너진다 —
 * `RegulatoryFlag`가 같은 이유로 짧은 라벨과 전체 문구를 나눠 두었다.
 */
export const APPLICABILITY_SHORT_LABEL: Readonly<Record<ApplicabilityState, string>> = {
  APPLICABLE: '',
  NOT_APPLICABLE: '규제 대상 아님',
  UNKNOWN: 'GT 미입력',
}

/**
 * 전체 문구 — `title`·`aria-label`로 나간다.
 *
 * `NOT_APPLICABLE`은 **`PRD §6.3`「공식 적용 대상 아님」의 원문 그대로**다. 문구를
 * 새로 쓰지 않는다(`AGENTS §3`). `UNKNOWN`은 `API_SPEC §1.6`
 * `CII_APPLICABILITY_UNKNOWN` 행의 사용자 메시지를 옮긴 것이다.
 */
export const APPLICABILITY_FULL_TEXT: Readonly<Record<ApplicabilityState, string>> = {
  APPLICABLE: '',
  NOT_APPLICABLE: '입력 선박은 공식 CII 적용 대상이 아닐 수 있습니다. 내부 분석용으로만 사용하세요.',
  UNKNOWN:
    '총톤수(GT)가 없어 공식 CII 적용 대상 여부를 판정할 수 없습니다. 선박 제원에 총톤수를 입력해 주세요.',
}
